Use empty smallvecs instead of `None` for `Target` and remove `NullableIndexMap`.

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
The following commits remove the usage of `Option` when referring to qargs in Rust and remove the `NullableIndexMap`.


### Details and comments
In an attempt to mimic the behavior of the `Target` in Python, there was an increased usage of `Option` in places where it was clearly not required. One of such happened to be the `gate_map` in which we kept an instance of `None` to indicate that an instruction was variadic. This was a bit problematic in rust as it wouldn't allow us to check for the presence of certain `qargs` in the `gate_map` due to ownership issues. Said constraints are what led to the inclusion of the `NullableIndexMap` in past iterations as it would allow the usage of `Option<&T>` for checking an index.

The solution to all of these issues is to, from now on, ditch the usage of `Option` when referring to `qargs` and use an empty slice (`[]`) internally instead. This also allows us to loosen up requirements for checking the target, as we can freely use `&[PhysicalQubit]` as a valid query. The following commits add these improvements while also modifying existing pieces of the code to leverage these new mechanics.

raynelfss committed:
Fix: Use empty smallvecs instead of `None`for `Target`
- In an attempt to mimic the behavior of the `Target` in Python, there was an increased usage of `Option` in places where it was clearly not required. One of such happened to be the `gate_map` in which we kept an instance of `None` to indicate that an instruction was variadic. This was a bit problematic in rust as it wouldn't allow us to check for the presence of certain `qargs` in the `gate_map` due to ownership issues. This is what led to the inclusion of the `NullableIndexMap` in past iterations as it would allow the usage of `Option<&T>` for checking an index.
The solution to all of these issues is to, from now on, ditch the usage of `Option` when referring to `qargs` and use an empty slice (`[]`) internally instead. This also allow us to loosen up requirements for checking the target, as we can freely use `&[PhysicalQubit]` as a valid query.
The following commits add these improvements while also modifying existing pieces of the code to leverage these new mechanics.

mtreinish commented on crates/accelerate/src/unitary_synthesis.rs:
```diff
@@ -519,15 +519,15 @@ fn get_2q_decomposers_from_target(
     let qubits: SmallVec<[PhysicalQubit; 2]> = SmallVec::from_buf(*qubits);
     let reverse_qubits: SmallVec<[PhysicalQubit; 2]> = qubits.iter().rev().copied().collect();
```
Do these even need to be a smallvec anymore? Like `qubits the input is already a slice, so we shouldn't need to do anything special here and maybe just do:

```suggestion
    let reverse_qubits: [PhysicalQubit; 2] = [qubits[1], qubits[0]];
```

mtreinish commented on crates/accelerate/src/consolidate_blocks.rs:
```diff
@@ -48,8 +48,9 @@ fn is_supported(
-            let physical_qargs = qargs.iter().map(|bit| PhysicalQubit(bit.0)).collect();
-            target.instruction_supported(name, Some(&physical_qargs))
+            let physical_qargs: Vec<PhysicalQubit> =
+                qargs.iter().map(|bit| PhysicalQubit(bit.0)).collect();
```
Not for this PR since this was already existing. But, I feel like we should be able to come up with a pattern here to avoid the allocation. At the end of the day we have a `&[u32]` in both cases, it's just the typing in the compiler we need to convert.

In the meantime I actually think a smallvec is probably more appropriate for places like this because it will be inlined most of the time.

raynelfss replied:
Fixed in 6a80a38

mtreinish commented on crates/accelerate/src/gates_in_basis.rs:
```diff
@@ -35,8 +34,8 @@ fn any_gate_missing_from_target(dag: &DAGCircuit, target: &Target) -> PyResult<b
     ) -> PyResult<bool> {
-        let qargs_mapped = SmallVec::from_iter(qargs.iter().map(|q| wire_map[q]));
-        if !target.instruction_supported(gate.op.name(), Some(&qargs_mapped)) {
+        let qargs_mapped: Vec<PhysicalQubit> = qargs.iter().map(|q| wire_map[q]).collect();
```
Same comment here I think using a smallvec here is fine. Ideally we shouldn't need to do a collection, but in the meantime I think using a smallvec is probably better performing. Although it's probably worth benchmarking `Vec` vs `SmallVec` here and the other places to see if there is a measurable difference.

raynelfss replied:
Fixed in 6a80a38. But didn't benchmark its impact.

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -296,24 +295,29 @@ impl Target {
+                            Some(qargs) => qargs.clone(),
+                            None => smallvec![],
+                        })
+                        .collect();
```
I realize this was pre-existing, but why do we need to collect here? Like I see why it's needed in the current because you're iterating over properties' keys and mutating properties. But I feel like this could be refactored to not need to mutate anything in properties. Maybe for a followup though.

raynelfss replied:
Solved in f3e88b1

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -368,21 +374,21 @@ impl Target {
         qargs: Option<Qargs>,
         properties: Option<InstructionProperties>,
     ) -> PyResult<()> {
+        let qargs_extract = qargs.as_ref().cloned().unwrap_or_default();
```
Why do we need to clone here? We didn't have that before

raynelfss replied:
This was a mistake in my end, this can probably be modified to be something along the lines of:
```suggestion
        let qargs_extract = qargs.as_ref().unwrap_or(&[]);
```

raynelfss replied:
Fixed in 017f657

mtreinish commented on crates/accelerate/src/target_transpiler/mod.rs:
```diff
@@ -1076,24 +1091,21 @@ impl Target {
         let mut qargs = qargs;
         if self.num_qubits.unwrap_or_default() == 0 || self.num_qubits.is_none() {
-            qargs = None;
+            qargs = [].as_slice();
```
Does `&[]` not work?

raynelfss replied:
For this case it should, there are just some cases in which it doesn't work as easily.

raynelfss replied:
Fixed in f3e88b1

mtreinish reviewed (commented):
Thanks for doing this, I'm very much looking forward to being able to use a slice directly instead of having to create smallvecs everywhere. Just a few inline comments to start.

mtreinish commented:
Oh I also just realized there might be implications with this change for inserting zero qubit gates into a target. Although, I don't think that is something supported by the target model more generally especially so it might not matter. Especially because our only zero qubit standard gate is `GlobalPhaseGate` which really isn't a thing for hardware (although aer could support it). But it might be worth figuring out what happens without this PR and see if anything changes with this PR.

jakelishman commented:
I was also going to comment about `None` and `Some(&[])` having different semantics (global/variadiac vs explicitly zero-arity), but I think in practice we might be able to infer which is meant.

That said, I also think there's trait implementations of `Equivalence` we could make so the `Option`-system work, but I haven't thought deeply about it.

raynelfss committed:
Lint: Fix formatting

raynelfss committed:
Apply suggestions from code review

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

raynelfss committed:
Fix: Use smallvec in changed collections.

raynelfss committed:
Fix: Avoid cloning when able, use `as_deref()` instead

raynelfss commented:
> @jakelishman: I was also going to comment about None and Some(&[]) having different semantics (global/variadiac vs explicitly zero-arity), but I think in practice we might be able to infer which is meant.

Is this a big concern? Right now testing doesn't break because it still relies on the Python side mapping for certain functions, which still keeps the `None`.

raynelfss committed:
Fix: Avoid collecting property keys during `add_instruction`.
- Use `&[]` instead of `[].as_slice()` when possible.

raynelfss committed:
Fix: Preserve `Option` as function argument,
- Re-arrange `add_instruction` to correct allocation issues.
- Re-arrange `operation_names_for_qargs` to skip conditions that need qargs to be a `Some` variant.
- Adapt codebasde to these changes.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into fix-target-none

raynelfss committed:
Docs: Add docstring for `Target::check_non_local_operation`.
- Fix errors during rebase.
